### ISSUE TITLE
Understanding the batching process for dataset like GigaSpeech

### ISSUE THREAD

**Issue by mrezasoltani** [lhotse-speech/lhotse#938]
Hi,

I am trying to understand the batching process for dataset like GigaSpeech. I have a very large telephone conversation (>100000 long records). I am playing with 5 records and try to understand the lhotse features for my dataset. 

I've adapted the GigaSpeech recipe for these 5 records and created the cuts from these records. 

This is how they looks like:
Record      # Supervisionsegment
Record_0	                 11 
Record_1	                 17
Record_2	                 83
Record_3	                 76
Record_4	                 46

For example, the first record has the following number of samples for each Supevisionsegment:

Supervisiosegment   # Samples
('Segment_0', 9200),
 ('Segment_1', 171200),
 ('Segment_2', 48960),
 ('Segment_3', 4640),
 ('Segment_4', 169120),
 ('Segment_5', 1440),
 ('Segment_6', 34400),
 ('Segment_7', 93280),
 ('Segment_8', 49440),
 ('Segment_9', 11520),
 ('Segment_10', 3600)

I am using the following codes to generate batches:

batch_size = 4
dataset = K2SpeechRecognitionDataset(input_strategy=AudioSamples())
sampler = DynamicBucketingSampler(
                            cuts, shuffle=False, max_cuts=batch_size,
                            max_duration=float('inf'), num_buckets=3)
dev_dloader = torch.utils.data.DataLoader(
    dataset,
    sampler=sampler,
    batch_size=None,
    num_workers=0)

My understanding from the sampler is that it should sample different cuts from these 5 records based on some method (e.g., DynamicBucketing). 
The above generates 2 batches like the followings:
batch_1['inputs'].shape = torch.Size([4  10875208])
batch_2['inputs'].shape = torch.Size([1  12935848])

From the above, it seems it uses the whole records to construct batches. I expect that the batches come from the Supervisionsegments simply by choosing randomly the Supervisionsegments, and collate them. I can do this by using cuts.trim_to_supervisions() inside of the DynamicBucketingSampler ( this results in 59 batches):

batch_1['inputs'].shape = torch.Size([4	171200])	
batch_2['inputs'].shape = torch.Size([4	142560])	
batch_3['inputs'].shape = torch.Size([4	48960]))
....

Does this make the process of generating batches slow? I could not find any example in the github to discuss about creating batches from a large dataset. 

Thanks for your hints

**Comment by pzelasko** [comment]
You got this correct, `cuts.trim_to_supervisions()` is the right approach here. It doesn't make batch generation slower, because it works on the metadata level. It effectively makes the load_audio function read only the required part of the underlying large audio file. The only case where it would be very slow is if you are importing Kaldi wav.scp and it has ffmpeg/sox pipes instead of plain paths to audio files.

**Comment by mrezasoltani** [comment]
Thanks for the answer. Also, I have two questions: 
1 - As you see in the above code, I set the max_duration=float('inf'). However, I get the following warning:
"UserWarning: Your 'buffer_size' setting of 10000 might be too low to satisfy a 'max_duration' of inf (given our best guess)."
Is this ok?

2 - As you have mentioned in your examples, one can set max_duration to some number for bucketing process. My question is this: when dataloader takes the sampler, does it try to create batches based on the max_duration value? if so, the batch size may different across batches?

Thanks

**Comment by pzelasko** [comment]
> Thanks for the answer. Also, I have two questions: 1 - As you see in the above code, I set the max_duration=float('inf'). However, I get the following warning: "UserWarning: Your 'buffer_size' setting of 10000 might be too low to satisfy a 'max_duration' of inf (given our best guess)." Is this ok?

This is OK, just keep in mind that the batches you will get from this may be wildly different in terms of the sequence length.

> 2 - As you have mentioned in your examples, one can set max_duration to some number for bucketing process. My question is this: when dataloader takes the sampler, does it try to create batches based on the max_duration value? if so, the batch size may be different across batches?

Yes, with bucketing sampler + max duration you will be getting batches with a different number of utterances ("batch size") but so that their total duration is kept close to max_duration value. This helps with maintaining better GPU utilization during training.

**Comment by mrezasoltani** [comment]
> > Thanks for the answer. Also, I have two questions: 1 - As you see in the above code, I set the max_duration=float('inf'). However, I get the following warning: "UserWarning: Your 'buffer_size' setting of 10000 might be too low to satisfy a 'max_duration' of inf (given our best guess)." Is this ok?
> 
> This is OK, just keep in mind that the batches you will get from this may be wildly different in terms of the sequence length.
> 
> > 2 - As you have mentioned in your examples, one can set max_duration to some number for bucketing process. My question is this: when dataloader takes the sampler, does it try to create batches based on the max_duration value? if so, the batch size may be different across batches?
> 
> Yes, with bucketing sampler + max duration you will be getting batches with a different number of utterances ("batch size") but so that their total duration is kept close to max_duration value. This helps with maintaining better GPU utilization during training.

Many thanks.
